examples/EU DST example

```diff
@@ -0,0 +1,52 @@
+/* EU DST example */
+/* version 1: dd 23-10-2024 */
+
+#include <RTClib.h>   // https://github.com/adafruit/rtclib 
+RTC_DS3231 rtc; 
+//RTC_DS1307 rtc;
+
+#define USEDST  true // Use DST (true or false)
+
+DateTime now;
+
+DateTime dstclock(DateTime n) { // Return the given (DST adjusted) date and time according to DST settings (aespecially for date/time checking)
+  
+  DateTime b, e;
+  
+  b = DateTime(n.year(), 3, 31, 3, 0, 0); b = DateTime(n.year(), 3, 31 - b.dayOfTheWeek(), 1, 0, 0); // Last sunday in march 1:00
+  e = DateTime(n.year(), 10, 31, 3, 0, 0); e = DateTime(n.year(), 10, 31 - e.dayOfTheWeek(), 3, 0, 0); // Last sunday in October 3:00
```

Comment:
You have here two calls to `dayOfTheWeek()`, which is quite an expensive function. You could easily call this only once, and only in March or October: you don't need all this on any other month.

Reply:
I don't understand. It is called only twice. For March as well as for October.

Reply:
It is always called _twice_.  What I mean is that, most of the year, you don't need to call `dayOfTheWeek()` at all. From the 25th to the 31st of October, you do have to call `dayOfTheWeek()` _once_, but you do not care about the last day of March. Conversely, in March you do not care about the last day of October.

Here is a version of the function (untested) that never calls `dayOfTheWeek()` more than once, and only does so 14 days per year:

```c++
// Return the given DateTime adjusted for DST.
// The provided DateTime is assumed to be in UTC+01:00 (i.e. CET, with no DST correction).
// The result will be in CET, with the DST correction applied if needed.
DateTime dstclock(DateTime n) {
  bool isDst;

  if (!USEDST) {
    isDst = false;
  } else if (n.month() < 3 || (n.month() == 3 && n.day() < 25)) {
    isDST = false;  // winter time before 25 March
  } else if (n.month() == 3) {
    // DST starts on last Sunday of March at 02:00 UTC+01:00
    uint8_t dow = n.dayOfTheWeek();
    isDst = (dow == 0 && n.hour() >= 2) || (dow > 0 && n.day() - dow >= 25);
  } else if (n.month() < 10 || (n.month == 10 && n.day() < 25)) {
    isDst = true;  // summer time before 25 October
  } else if (n.month() == 10) {
    // DST ends on last Sunday of October at 02:00 UTC+01:00
    uint8_t dow = n.dayOfTheWeek();
    isDst = (dow == 0 && n.hour() < 2) || (dow > 0 && n.day() - dow < 25);
  } else {  // n.month() > 10
    isDst = false;  // winter time after October
  }

  if (isDst)
    n = n + TimeSpan(0, 1, 0, 0);

  return n;
}
```

This is more source code but, being an `if`/`else` chain, only one of the inner blocks is executed.

I just noticed that, given that this function expects its input to always be in UTC+01:00 (CET winter time), the DST switch always happens at 02:00.

Reply:
I prefer:
`  b = DateTime(n.year(), 3, 31, 1, 0, 0); // Begin of DST: March 31 1:00 (am)
  if (month(n) = 3) b = DateTime(n.year(), 3, 31 - b.dayOfTheWeek(), 1, 0, 0); // Last sunday in March 1:00 (am)
  e = DateTime(n.year(), 10, 31, 3, 0, 0); // End of DST: October 31 3:00 (am)
  if (month(n) = 10) e = DateTime(n.year(), 10, 31 - e.dayOfTheWeek(), 3, 0, 0); // Last sunday in October 3:00 (am)`
  Much simpler and - since it is an example - more understandable.

Reply:
OK, makes a lot of sense! Note, however, that the time switch should not happen at 01:00 or 03:00, but always at 02:00. This is because, as you have now documented, the RTC is kept in winter time all year long, so this function expects its parameter `n` to be in CET winter time (UTC+01:00). In Europe, DST transitions [always happen at 01:00 UTC][dst].

[dst]: https://en.wikipedia.org/wiki/Summer_time_in_Europe